V2.0 jazzcat cache override

noahadams opened the pull request:
Adds a "cacheOverrideTime" parameter to Jazzcat.combineScripts()'s options object that allows one to override the behaviour of isStale when loading scripts.

noahadams committed:
add unit tests for httpCache.utils.isStale

noahadams committed:
initial implementation of an option to override caching time for jazzcat

noahadams committed:
rename some things, add a test suite for the overrides

noahadams committed:
test setting and unsetting of override

jansepar committed:
Merge remote-tracking branch 'origin/v2.0' into v2.0-jazzcat-cache-override

Conflicts:
	tests/jazzcat.html

jansepar committed:
Merge branch 'v2.0-jazzcat-cache-override' of github.com:mobify/mobifyjs into v2.0-jazzcat-cache-override

Conflicts:
	tests/jazzcat.html

noahadams committed:
make the cache override an option to isStale as applied by the httpCache.load function, stub out load tests

noahadams committed:
Merge remote-tracking branch 'origin/v2.0' into v2.0-jazzcat-cache-override

Conflicts:
	src/jazzcat.js
	tests/jazzcat.html

noahadams committed:
re-order tests, put httpcache stuff first

noahadams committed:
more resource fixtures

noahadams committed:
rename test

noahadams committed:
make loader generation code sensitive to cache override option

noahadams committed:
integrationish test for cache override

noahadams committed:
fix up test to use outerHTML

jansepar commented on tests/jazzcat.html:
```diff
@@ -215,6 +361,60 @@
+
+                    // verify that things we wanted to load from cache did and other things did not
+
+                    var enabledDocumentString = Capture.enable(Utils.outerHTML(doc.documentElement));
```
Not a huge deal but curious why you used Capture.enable here, instead of just leaving `x-` inside of `expectedHTML`?

noahadams replied:
I wanted to see that the correct bootloader was generated in the pass through enable. We could also unit test the functions this depends on.

jansepar replied:
Ohhhh okay, I was thinking that you use enable to remove the `x-`. For most of my tests I just used `insertLoadersIntoHTMLString` to test the bootloader insertion, but I guess it's good to have at least one test using `enable` to ensure that method doesn't regress.

jansepar commented on src/jazzcat.js:
```diff
@@ -270,7 +287,12 @@ define(["utils", "capture"], function(Utils, Capture) {
      */
+
     Jazzcat.combineScripts = function(scripts, doc, options) {
+        if (options && options.cacheOverrideTime !== undefined) {
```
Can we switch this to: `if (options && options.cacheOverrideTime) {`.

I believe it's more consistent with the style of our other `if` statements, and it's more legible.

noahadams replied:
but what if you want to override it to 0? We could do some more serious validation as well, like, check the type and check that it's in an integer in [0, infty)

jansepar replied:
Yeah I guess that should be possible (although not likely). Fair enough!

jansepar commented on src/jazzcat.js:
```diff
@@ -396,7 +419,11 @@ define(["utils", "capture"], function(Utils, Capture) {
+          "Jazzcat.httpCache.load(" + JSON.stringify(httpCache.options) + ");" :
+          "Jazzcat.httpCache.load();" );
+
+        //loadFromCacheScript.innerHTML = "Jazzcat.httpCache.load();"
```
Nuke this comment from orbit :)

jansepar commented on src/jazzcat.js:
```diff
@@ -270,7 +287,12 @@ define(["utils", "capture"], function(Utils, Capture) {
      * bootloader script is inserted by the overriden `Capture.enabled` function.
      */
+
     Jazzcat.combineScripts = function(scripts, doc, options) {
```
Could you throw into the comments above here what the options are that you can pass in?

jansepar commented on src/jazzcat.js:
```diff
@@ -58,9 +60,14 @@ define(["utils", "capture"], function(Utils, Capture) {
         var key;
+        var staleOptions;
+
+        if (options && options.overrideTime) {
```
Should probably check for undefined here if we are being consistent with `combineScripts`.

noahadams replied:
touché

jansepar commented on src/jazzcat.js:
```diff
@@ -176,13 +183,22 @@ define(["utils", "capture"], function(Utils, Capture) {
+        // response is less than the override, cacheOverrideTime is in minutes, 
+        // turn it off by setting it to false
+        if (options && (overrideTime = options.overrideTime) &&
+          (date = Date.parse(headers.date))) {
```
Why throw `date = Date.parse(headers.date)` into the `if`? What's the condition we are checking that might opt us out here?

noahadams replied:
the assignment in the if was a brevity thing for validating the parsing, but this can just get moved up.

jansepar commented on src/jazzcat.js:
```diff
@@ -176,13 +183,22 @@ define(["utils", "capture"], function(Utils, Capture) {
-        // If `max-age` and `date` are present, and no other no other cache
+        // If `max-age` and `date` are present, and no other cache
         // directives exist, then we are stale if we are older.
         if (cacheControl && (date = Date.parse(headers.date))) {
```
You already create this `date` variable, don't need to parse again.

noahadams committed:
PR comment answers

- Add default 10 minute caching to not conflict with server's default 10-minute caching
- Comment added parameter
- Cleanups around guards and some whitespace

noahadams committed:
tweak tests, add back a guard on date in isStale

jansepar commented on src/jazzcat.js:
```diff
@@ -269,8 +290,21 @@ define(["utils", "capture"], function(Utils, Capture) {
     Jazzcat.combineScripts = function(scripts, doc, options) {
+        if (options && options.cacheOverrideTime !== undefined) {
+            Utils.extend(httpCache.options,
+              {overrideTime: options.cacheOverrideTime});
```
Don't we want to default the overrideTime to 10? Currently, this will cause a Jazzcat request (which will hit cache for 10 min), but I thought we decided to prevent that?

noahadams replied:
The default of ten minutes is implemented directly in the isStale function, that's not an override.

jansepar committed:
Updating the mobify-custom.js.example to help test this PR

jansepar commented on src/jazzcat.js:
```diff
@@ -173,18 +180,31 @@ define(["utils", "capture"], function(Utils, Capture) {
+
+        // Fresh if less than 10 minutes old
+        if (date && (now < date + 600 * 1000)) {
+            return false;
```
Ah, I never saw this because I was looking at the next `if` statement. We could save ourself from running this logic twice if we just set `overrideTime = 10`. Or maybe something like this:

`cacheTime = options.overrideTime || 10`

noahadams replied:
idk, I sort of like how the cascade of early-returning ifs that's already here reads: 

Fresh if newer than 10 minutes OR newer than override if present OR newer than http cache time if present
Otherwise, stale.

jansepar committed:
Committing build files

jansepar commented:
:+1:
